rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,234 @@
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Alternate `pkgs/unit` structure
+
+- Use a flat directory, e.g. `pkgs.hello` would be in `pkgs/unit/hello`.
+  - Good because it's simpler, both for the user and for the code
+  - Bad because the GitHub web interface only renders the first 1 000 entries (and we have about 10 000 that benefit from this transition, even given the restrictions)
+  - Bad because it makes `git` and file listings slower
+- Use `substring 0 3 name` or `substring 0 4 name`. This was not done because it still leads to directories in `pkgs/unit` containing more than 1 000 entries, leading to the same problems.
+- Use multi-level structure, like a 2-level 2-prefix structure where `hello` is in `pkgs/unit/he/ll/hello`,
+  if packages are less than 4 characters long, we will it out with `-`, e.g. `z` is in `pkgs/unit/z-/--/z`.
+  This is not great because it's more complicated, longer to type and it would improve performance only marginally.
+- Use a dynamic structure where directories are rebalanced when they have too many entries.
+  E.g. `pkgs.foobar` could be in `pkgs/unit/f/foobar` initially.
+  But when there's more than 1 000 packages starting with `f`, all packages starting with `f` are distributed under 2-letter prefixes, moving `foobar` to `pkgs/unit/fo/foobar`.
+  This is not great because it's very complex to determine which directory to put a package in, making it bad for contributors.
+
+## Alternate `pkg-fun.nix` filename
+
+- `default.nix`:
+  - Bad because it doesn't have its main benefits here:
+    - Removing the need to specify the file name in expressions, but this does not apply because this file will be imported automatically by the code that replaces definitions from `all-packages.nix`.
+    - Removing the need to specify the file name on the command line, but this does not apply because a package function must be imported into an expression before it can be used, making `nix build -f pkgs/unit/hell/hello` equally broken regardless of file name.
+  - Not using `default.nix` frees up `default.nix` for a short expression that is actually buildable, e.g. `(import ../.. {}).hello`, although we don't yet have a use case for this that isn't covered by `nix-build ../.. -A $attrname`.
+  - Bad because using `default.nix` would tempt users to invoke `nix-build .` whereas making package functions auto-callable is a known anti-pattern as it duplicates the defaults.
+  - Good because `default.nix` is already a convention most people are used to
+- `package.nix`/`pkg.nix`: Bad, because it makes the migration to a non-function form of overridable packages harder in the future.
+
+## Alternate `pkgs/unit` location
+
+- Use `unit` (at the Nixpkgs root) instead of `pkgs/unit`.
+  This is future proof in case we want to make the directory structure more general purpose, but this is out of scope
+- Other name proposals were deemed worse: `pkg`, `component`, `part`, `mod`, `comp`
```

Comment:
I would like to nominate `auto`, to me this is more future-proof than `unit` so we can include multiple packages in one directory

Reply:
> so we can include multiple packages in one directory

How do you mean that?

Reply:
For example, we currently have `element-desktop` and `element-web` that share the same `pin.json` (or `generic.nix` in some other cases), in the future we might want to allow that to be defined in `pkgs/unit`, but I'm not quite sure the name `unit` fits the behavior of having multiple packages under the same directory, and `auto` is more future-proof in that case IMO

Reply:
I see, yeah so this has been discussed before. I believe in the team the consensus was that multiple versions and variants should eventually end up in the same `pkgs.element` attribute, once we have the design for that figured out. In that case, all those variants would naturally be together in the `pkgs/unit/el/element` directory.

That's also kind of why `unit` was chosen, we talked about [explaining it better](https://github.com/nixpkgs-architecture/simple-package-paths/issues/16) before, something like:

> unit: A collection of standardized files related to the same software component

Reply:
that makes sense to me, thanks for the explanation

Reply:
I don't dislike "auto" but I'd like to note that the name implies the package is *automatically* called. This means that packages with custom arguments would not fit in there anymore, unlike with the current proposal. Maybe they'd be put into "pkgs/manual" instead then? (which then opens up to confusion with manual as in documentation, etc.)